packages/flutter_tools/lib/src/commands/upgrade.dart

```diff
@@ -20,42 +20,38 @@ class UpgradeCommand extends FlutterCommand {
   final String description = 'Upgrade your copy of Flutter.';
 
   @override
-  Validator projectRootValidator = () => true;
+  bool get requiresProjectRoot => false;
 
   @override
   Future<int> runInProject() async {
-    printStatus(FlutterVersion.getVersion(ArtifactStore.flutterRoot).toString());
-
     try {
       runCheckedSync(<String>[
         'git', 'rev-parse', '@{u}'
       ], workingDirectory: ArtifactStore.flutterRoot);
     } catch (e) {
-      printError('Unable to upgrade Flutter. No upstream repository configured for Flutter.');
+      printError('Unable to upgrade Flutter: no upstream repository configured.');
       return 1;
     }
 
-    printStatus('');
     printStatus('Upgrading Flutter...');
 
-    int code = await runCommandAndStreamOutput(<String>[
-      'git', 'pull', '--ff-only'
-    ], workingDirectory: ArtifactStore.flutterRoot);
+    int code = await runCommandAndStreamOutput(
+      <String>['git', 'pull', '--ff-only'],
+      workingDirectory: ArtifactStore.flutterRoot,
+      mapFunction: (String line) => matchesGitLine(line) ? null : line
+    );
 
     if (code != 0)
       return code;
 
-    // Causes us to update our locally cached packages.
-    printStatus('');
-    code = await runCommandAndStreamOutput(<String>[
-      'bin/flutter', '--version'
-    ], workingDirectory: ArtifactStore.flutterRoot);
-
-    if (code != 0)
-      return code;
+    // Check for and download any engine and pkg/ updates.
+    if (!cache.isUpToDate()) {
+      printStatus('');
+      await cache.updateAll();
```

Comment:
This can get us into an inconsistent state.  Suppose we add a new package or other cache entry.  Now we're using the old flutter_tools code to download artifacts specified by a newer version of the cache.  We need to switch to using the new code at this point to guarantee consistency.